teps/0142-enable-step-reusability.md

```diff
@@ -0,0 +1,1071 @@
+
+<table>
+<tr>
+<td>
+
+```yaml
+apiVersion: tekton.dev/v1alpha1
+kind: StepAction
+metadata:
+  name: step1
+spec:
+  image: alpine
+  script: |
+    echo "I am expendable"
+```
+</td>
+<td rowspan=2>
+
+```yaml
+apiVersion: tekton.dev/v1
+kind: Task
+metadata:
+  name: ci-t
+spec:
+  steps:
+    - name: s1
+      stepRef:
+        name: step1
+    - name: s2
+      stepRef:
+        name: step2
+```
+</td>
+</tr>
+<tr>
+<td>
+
+```yaml
+apiVersion: tekton.dev/v1alpha1
+kind: StepAction
+metadata:
+  name: step2
+spec:
+  image: alpine
+  script: |
+    echo "I am special"
+```
+</td>
+</tr>
+</table>
+
+A `Task` can refer to the `StepAction` from the cluster or fetch it from remote sources via a resolver. 
+
+<table>
+<tr>
+<td>StepAction from cluster</td>
+<td>StepAction from Git repository</td>
+</tr>
+<tr>
+<td>
+
+```yaml
+apiVersion: tekton.dev/v1
+kind: Task
+metadata:
+  name: ci-t
+spec:
+  steps:
+    - name: s1
+      stepRef:
+        name: step1
+```
+</td>
+<td>
+
+```yaml
+apiVersion: tekton.dev/v1
+kind: Task
+metadata:
+  name: ci-t
+spec:
+  steps:
+    - name: s1
+      stepRef:
+        resolver: git
+        params:
+          - name: url
+            value: https://...
+```
+</td>
+</tr>
+</table>
+
+<table>
+<tr>
+<td>Example StepAction</td>
+<td>Example Task with orchestration in Steps</td>
+</tr>
+<tr>
+<td>
+
+```yaml
+apiVersion: tekton.dev/v1alpha1
+kind: StepAction
+metadata:
+  name: step1
+spec:
+  image: alpine
+  script: |
+    echo "I am expendable"
+
+```
+</td>
+<td rowspan=2>
+
+```yaml
+apiVersion: tekton.dev/v1
+kind: Task
+metadata:
+  name: ci-t
+spec:
+  steps:
+    - name: s1
+      stepRef:
+        name: step1
+      onError: continue
+    - name: s2
+      stepRef:
+        name: step2
+      computeResources:
+        requests:
+          memory: 1Gi
+          cpu: 500m
+        limits:
+          memory: 2Gi
+          cpu: 800m
+```
+</td>
+</tr>
+<tr>
+<td>
+
+```yaml
+apiVersion: tekton.dev/v1alpha1
+kind: StepAction
+metadata:
+  name: step2
+spec:
+  image: alpine
+  script: |
+    echo "I am special"
+  securityContext:
+    privileged: true
+```
+</td>
+</tr>
+</table>
+
+#### Params
+If a `StepAction` needs `Parameters`, they need to be declared so that they can be provided by the `Task` author. Without declaring `Parameters`, the `Task` author would have had to go through the `StepAction` specification to decipher what is needed. A `Task` should declare the `Parameters` that the underlying `Step` needs so that they can be provided by the `TaskRun` or the `Pipeline`. 
+<table>
+<tr>
+<td>
+
+```yaml
+# This is how a step author would 
+# declare the params that it needs.
+apiVersion: tekton.dev/v1alpha1
+kind: StepAction
+metadata:
+  name: myStep
+spec:
+  params:
+    - name: foo
+      type: string
+      description: “This is a param”
+      default: bar
+  image: alpine
+  script: |
+    echo $(params.foo)
```

Comment:
Thanks for bringing that up. I didn't really know that. May be because we use examples in our codebase that are filled with using params directly inside scripts. A lot of my knowledge on the uses is from those examples. I wonder if others are in the same boat as well. 

I think we should really have guidelines/best practices doc that lays all this out. 

Will fix!

Reply:
@chitrangpatel we have guidance in the [catalog docs](https://github.com/tektoncd/catalog/blob/main/recommendations.md#dont-use-interpolation-in-scripts-or-string-arguments) and now [this TEP](https://github.com/tektoncd/community/pull/1077) too. 

I know we have this still in many examples, which we should fix (even though we still need to keep it somewhere for testing purposes), but I would like to avoid adding more examples. Could you please fix the example in the TEP?

Reply:
Fixed!